src/Components/Components/src/NavigationManagerExtensions.cs

```diff
@@ -649,14 +656,30 @@ private static string GetUriWithRemovedQueryParameter(string uri, string name)
             }
         }
 
-        return newQueryStringBuilder.UriWithQueryString;
+        return newQueryStringBuilder.GetUriWithQueryString();
     }
 
     private static string GetUriWithAppendedQueryParameters(
         string uriWithoutQueryString,
         IReadOnlyDictionary<string, object?> parameters)
     {
-        var builder = new QueryStringBuilder(uriWithoutQueryString);
+        ReadOnlySpan<char> uriWithoutQueryStringAndHash;
+        ReadOnlySpan<char> hash;
+
+        var hashStartIndex = uriWithoutQueryString.IndexOf('#');
+
+        if (hashStartIndex < 0)
+        {
+            uriWithoutQueryStringAndHash = uriWithoutQueryString;
+            hash = "";
+        }
+        else
+        {
+            uriWithoutQueryStringAndHash = uriWithoutQueryString.AsSpan(0, hashStartIndex);
+            hash = uriWithoutQueryString.AsSpan(hashStartIndex);
+        }
+
+        var builder = new QueryStringBuilder(uriWithoutQueryStringAndHash, hash: hash.ToString());
```

Comment:
If we call `hash.ToString()` here, it will allocate a string which makes the use of `ReadOnlySpan<char>` redundant. Perhaps it's unavoidable, and if so, sorry for the misleading suggestion about `ReadOnlySpan<char>`!

However, one other possibility: does it work to *not* pass `hash` to the `QueryStringBuilder` constructor at all? Instead, there could be a new method on `QueryStringBuilder` like `AppendHash(ReadOnlySpan<char> hash)` which gets called just before `GetUriWithQueryString`. That also avoids another allocation (the one caused by the concatenation `_builder.ToString() + _hash`). Does that work?

I know it might seem fussy to be doing all these tricks to avoid a couple of allocations! It's only for low-level APIs that might be called hundreds of times per request that we tend to focus so much on it.

Reply:
That is a great point to raise and would absolutely be something we would need to deal with if `QueryStringBuilder` was public API. But since in this case it's private API (and especially because it's scoped to this one file, so we don't even have to reason about other usages across the codebase), we don't have to worry too hard about incorrect usage patterns.

If you want to add a check that the methods aren't called in the wrong order that's fine, but TBH I'd be completely fine with not doing that and simply relying on the caller doing things correctly since we know that's just us. If we ever get this wrong in the future, we'll know about it because we'll get incorrect results.